subsys/suit/platform/sdfw/src/suit_plat_ipuc.c

```diff
@@ -45,49 +48,164 @@ static struct zcbor_string *ipuc_component_id_from_handle(suit_component_t handl
 	return component_id;
 }
 
-int suit_plat_ipuc_revoke(suit_component_t handle)
+static suit_plat_ipuc_entry_t *ipuc_entry_from_handle(suit_component_t handle)
 {
 	struct zcbor_string *component_id = ipuc_component_id_from_handle(handle);
 
 	if (component_id == NULL) {
-		return SUIT_ERR_UNSUPPORTED_COMPONENT_ID;
+		return NULL;
 	}
 
-	k_mutex_lock(&ipuc_mutex, K_FOREVER);
 	for (size_t i = 0; i < CONFIG_SUIT_IPUC_SIZE; i++) {
 		suit_plat_ipuc_entry_t *p_entry = &components[i];
 
 		if (suit_compare_zcbor_strings(component_id, &p_entry->component_id)) {
+			return p_entry;
+		}
+	}
 
-#ifdef CONFIG_SUIT_LOG_LEVEL_INF
-			intptr_t slot_address = 0;
-			size_t slot_size = 0;
+	return NULL;
+}
 
-			if (suit_plat_decode_address_size(&p_entry->component_id, &slot_address,
-							  &slot_size) == SUIT_PLAT_SUCCESS) {
-				LOG_INF("Revoking IPUC, address: %p, size: %d bytes",
-					(void *)slot_address, slot_size);
-			}
-#endif
-			p_entry->component_id.value = NULL;
-			p_entry->component_id.len = 0;
-			p_entry->write_offset = 0;
-			break;
+suit_plat_err_t suit_plat_ipuc_write(suit_component_t handle, size_t offset, uintptr_t buffer,
+				     size_t chunk_size, bool last_chunk)
+{
+	suit_plat_err_t plat_ret = SUIT_PLAT_SUCCESS;
+	struct stream_sink ipuc_sink;
+	bool sink_allocated = false;
+
+	k_mutex_lock(&ipuc_mutex, K_FOREVER);
+	suit_plat_ipuc_entry_t *ipuc_entry = ipuc_entry_from_handle(handle);
+
+	if (ipuc_entry == NULL) {
+		plat_ret = SUIT_PLAT_ERR_NOT_FOUND;
+	}
+
+	if (plat_ret == SUIT_PLAT_SUCCESS) {
+
+		if (offset == 0) {
+			ipuc_entry->write_offset = 0;
+			ipuc_entry->last_chunk_stored = false;
+		}
+
+		if (offset != ipuc_entry->write_offset || ipuc_entry->last_chunk_stored) {
+			plat_ret = SUIT_PLAT_ERR_INCORRECT_STATE;
+		}
+	}
+
+	if (plat_ret == SUIT_PLAT_SUCCESS) {
+		plat_ret = suit_sink_select(handle, &ipuc_sink);
+		if (plat_ret != SUIT_PLAT_SUCCESS) {
+			plat_ret = SUIT_PLAT_ERR_IO;
+		} else {
+			sink_allocated = true;
+		}
+	}
+
+	if (plat_ret == SUIT_PLAT_SUCCESS) {
+		if (ipuc_sink.seek == NULL) {
```

Comment:
Could you add a note here, why `seek` operation is strictly necessary for IPUC to work (I guess it is required so the sink can seek into the first address that can be used for SDFW updates)?

Reply:
Ahh.. now I see it :+1: